src/style-spec/expression/definitions/comparison.js

```diff
@@ -0,0 +1,170 @@
+    } else {
+        // ordering operator
+        return type.kind === 'string' ||
+            type.kind === 'number' ||
+            type.kind === 'value';
+    }
+}
+
+
+function eq(ctx, a, b) { return a === b; }
+function neq(ctx, a, b) { return a !== b; }
+function lt(ctx, a, b) { return a < b; }
+function gt(ctx, a, b) { return a > b; }
+function lteq(ctx, a, b) { return a <= b; }
+function gteq(ctx, a, b) { return a >= b; }
+
+function eqCollate(ctx, a, b, c) { return c.compare(a, b) === 0; }
+function neqCollate(ctx, a, b, c) { return !eqCollate(ctx, a, b, c); }
+function ltCollate(ctx, a, b, c) { return c.compare(a, b) < 0; }
+function gtCollate(ctx, a, b, c) { return c.compare(a, b) > 0; }
+function lteqCollate(ctx, a, b, c) { return c.compare(a, b) <= 0; }
+function gteqCollate(ctx, a, b, c) { return c.compare(a, b) >= 0; }
+
+/**
+ * Special form for comparison operators, implementing the signatures:
+ * - (T, T, ?Collator) => boolean
+ * - (T, value, ?Collator) => boolean
+ * - (value, T, ?Collator) => boolean
+ *
+ * For inequalities, T must be either value, string, or number. For ==/!=, it
+ * can also be boolean or null.
+ *
+ * Equality semantics are equivalent to Javascript's strict equality (===/!==)
+ * -- i.e., when the arguments' types don't match, == evaluates to false, != to
+ * true.
+ *
+ * When types don't match in an ordering comparison, a runtime error is thrown.
+ *
+ * @private
+ */
+function makeComparison(op: ComparisonOperator, compareBasic, compareWithCollator) {
+    return class Comparison implements Expression {
+        type: Type;
+        lhs: Expression;
+        rhs: Expression;
+        collator: ?Expression;
+        needsRuntimeTypeCheck: boolean;
+
+        constructor(lhs: Expression, rhs: Expression, collator: ?Expression) {
+            this.type = BooleanType;
+            this.lhs = lhs;
+            this.rhs = rhs;
+            this.collator = collator;
+            this.needsRuntimeTypeCheck = op !== '==' && op !== '!=' &&
+                lhs.type.kind === 'value' && rhs.type.kind === 'value';
+        }
+
+        static parse(args: Array<mixed>, context: ParsingContext): ?Expression {
+            if (args.length !== 3 && args.length !== 4)
+                return context.error(`Expected two or three arguments.`);
+
+            const op: ComparisonOperator = (args[0]: any);
+
+            let lhs = context.parse(args[1], 1, ValueType);
+            if (!lhs) return null;
+            if (!isComparableType(op, lhs.type)) {
+                return context.concat(1).error(`"${op}" comparisons are not supported for type '${toString(lhs.type)}'.`);
+            }
+            let rhs = context.parse(args[2], 2, ValueType);
+            if (!rhs) return null;
+            if (!isComparableType(op, rhs.type)) {
+                return context.concat(2).error(`"${op}" comparisons are not supported for type '${toString(rhs.type)}'.`);
+            }
+
+            if (!(
```

Comment:
I find it more difficult to understand this condition than if the `!` was distributed.